codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsProtocolUtils.java

```diff
@@ -162,4 +172,91 @@ public static void handleDecodeError(GoWriter writer, String returnExtras) {
+            Optional<XmlAttributeTrait> xmlAttributeTrait = memberShape.getTrait(XmlAttributeTrait.class);
+            if (xmlAttributeTrait.isPresent()) {
+                writeSafeMemberAccessor(context, memberShape, inputSrc, (operand) -> {
+                    writer.write("$L = append($L, smithyxml.NewAttribute($S, string(*$L)))",
+                            attrName, attrName,
+                            getSerializedXMLMemberName(memberShape), operand);
+                });
+            }
+        });
+
+        writer.openBlock("$L := smithyxml.StartElement{ ", "}", dst, () -> {
+            writer.openBlock("Name:smithyxml.Name{", "},", () -> {
+                writer.write("Local: $S,", getSerializedXMLShapeName(context, shape));
+            });
+            writer.write("Attr : $L,", attrName);
+        });
+    }
+
+
+    /**
+     * getSerializedXMLMemberName returns a xml member name used for serializing. If a member shape has
+     * XML name trait, xml name would be given precedence over member name.
+     *
+     * @param memberShape is the member shape for which serializer name is queried.
+     * @return name of a xml member shape used by serializers
+     */
+    private static String getSerializedXMLMemberName(MemberShape memberShape) {
+        Optional<XmlNameTrait> xmlNameTrait = memberShape.getTrait(XmlNameTrait.class);
+        return xmlNameTrait.isPresent() ? xmlNameTrait.get().getValue() : memberShape.getMemberName();
+    }
+
+    /**
+     * getSerializedXMLShapeName returns a xml shape name used for serializing. If a member shape
+     * has xml name trait, xml name would be given precedence over member name.
+     * This correctly handles renamed shapes, and returns the original shape name.
+     *
+     * @param context is the generation context for which
+     * @param shape   is the Shape for which serializer name is queried.
+     * @return name of a xml member shape used by serializers.
+     */
+    private static String getSerializedXMLShapeName(GenerationContext context, Shape shape) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol shapeSymbol = symbolProvider.toSymbol(shape);
+        String shapeName = shapeSymbol.getName();
+
+        // check if synthetic cloned shape
+        Optional<SyntheticClone> clone = shape.getTrait(SyntheticClone.class);
+        if (clone.isPresent()) {
+            SyntheticClone cl = clone.get();
+            shapeName = cl.getArchetype().getName();
+        }
+
+        // check if shape is member shape
+        Optional<MemberShape> member = shape.asMemberShape();
+        if (member.isPresent()) {
+            MemberShape memberShape = member.get();
+            return getSerializedXMLMemberName(memberShape);
+        }
+
+        Optional<XmlNameTrait> xmlNameTrait = shape.getTrait(XmlNameTrait.class);
+        return xmlNameTrait.isPresent() ? xmlNameTrait.get().getValue() : shapeName;
```

Comment:
Something like the following would be better:

`shape.getTrait(XmlNameTrait.class).map(XmlNameTrait::getValue).orElse(shapeName)`